File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        
        TreeMap<Integer, Integer> tree = new TreeMap<>();
        StringTokenizer st;
        
        for(int i=0; i<operations.length; i++){
            st = new StringTokenizer(operations[i]);
            String input = st.nextToken();
            int num = Integer.parseInt(st.nextToken());
            
            if(input.equals("I")){
                tree.put(num, tree.getOrDefault(num, 0)+1);
            }else{
                if(tree.isEmpty()) continue;
                if(num == 1){
                    if(tree.get(tree.lastKey()) == 1){
                        tree.remove(tree.lastKey());
                    }else{
                        tree.put(tree.lastKey(), tree.get(tree.lastKey())-1);
                    }
                }else{
                    if(tree.get(tree.firstKey()) == 1){
                        tree.remove(tree.firstKey());
                    }else{
                        tree.put(tree.firstKey(), tree.get(tree.firstKey())-1);
                    }
                }
            }
        }
        
        int answer[] = tree.isEmpty()? new int [] {0,0}: new int[] {tree.lastKey(), tree.firstKey()};
        
        return answer;
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.js
function solution(operations) {
    var answer = [];
    let que = [];
    
    for(let i=0; i<operations.length; i++){
        if(operations[i].charAt(0) == 'I'){
            que.push(operations[i].slice(2)*1);
        }else if(operations[i] == "D 1"){
            que.pop();
        }else{
            que.shift();
        }
        
        que.sort((a,b)=>a-b);
    }
    
    answer = que.length? [que[que.length-1],que[0]] : [0, 0];
    
    return answer;
}